cudax/include/cuda/experimental/__async/sender/intrusive_queue.cuh

```diff
@@ -0,0 +1,319 @@
+
+#include <cuda/experimental/__async/sender/prologue.cuh>
+
+namespace cuda::experimental::__async
+{
+template <auto _Next>
+class _CCCL_TYPE_VISIBILITY_DEFAULT __intrusive_queue;
+
+template <class _Item, _Item* _Item::* _Next>
+class _CCCL_TYPE_VISIBILITY_DEFAULT __intrusive_queue<_Next>
+{
+public:
+  _CUDAX_DEFAULTED_API __intrusive_queue() noexcept = default;
+
+  _CUDAX_API __intrusive_queue(__intrusive_queue&& __other) noexcept
+      : __head_(_CUDA_VSTD::exchange(__other.__head_, nullptr))
+      , __tail_(_CUDA_VSTD::exchange(__other.__tail_, nullptr))
+  {}
+
+  _CUDAX_API __intrusive_queue(_Item* __head, _Item* __tail) noexcept
+      : __head_(__head)
+      , __tail_(__tail)
+  {}
+
+  _CUDAX_API auto operator=(__intrusive_queue __other) noexcept -> __intrusive_queue&
+  {
+    _CUDA_VSTD::swap(__head_, __other.__head_);
+    _CUDA_VSTD::swap(__tail_, __other.__tail_);
+    return *this;
+  }
+
+  _CUDAX_API ~__intrusive_queue()
+  {
+    _CCCL_ASSERT(empty(), "");
+  }
+
+  _CUDAX_API static auto make_reversed(_Item* __list) noexcept -> __intrusive_queue
+  {
+    _Item* __new_head = nullptr;
+    _Item* __new_tail = __list;
+
+    while (__list != nullptr)
+    {
+      _Item* __next  = __list->*_Next;
+      __list->*_Next = __new_head;
+      __new_head     = __list;
+      __list         = __next;
+    }
+
+    __intrusive_queue __result;
+    __result.__head_ = __new_head;
+    __result.__tail_ = __new_tail;
+    return __result;
+  }
+
+  _CUDAX_API static auto make(_Item* __list) noexcept -> __intrusive_queue
+  {
+    __intrusive_queue __result{};
+    __result.__head_ = __list;
+    __result.__tail_ = __list;
+    if (__list == nullptr)
+    {
+      return __result;
+    }
+    while (__result.__tail_->*_Next != nullptr)
+    {
+      __result.__tail_ = __result.__tail_->*_Next;
+    }
+    return __result;
+  }
+
+  [[nodiscard]]
+  _CUDAX_API auto empty() const noexcept -> bool
+  {
+    return __head_ == nullptr;
+  }
+
+  _CUDAX_API void clear() noexcept
+  {
+    __head_ = nullptr;
+    __tail_ = nullptr;
+  }
+
+  [[nodiscard]]
+  _CUDAX_API auto pop_front() noexcept -> _Item*
+  {
+    _CCCL_ASSERT(!empty(), "");
+    _Item* __item = _CUDA_VSTD::exchange(__head_, __head_->*_Next);
+    // This should test if __head_ == nullptr, but due to a bug in
+    // nvc++'s optimization, `__head_` isn't assigned until later.
+    // Filed as NVBug#3952534.
+    if (__item->*_Next == nullptr)
+    {
+      __tail_ = nullptr;
+    }
+    return __item;
+  }
+
+  _CUDAX_API void push_front(_Item* __item) noexcept
+  {
+    _CCCL_ASSERT(__item != nullptr, "");
+    __item->*_Next = __head_;
+    __head_        = __item;
+    if (__tail_ == nullptr)
+    {
+      __tail_ = __item;
+    }
+  }
+
+  _CUDAX_API void push_back(_Item* __item) noexcept
+  {
+    _CCCL_ASSERT(__item != nullptr, "");
+    __item->*_Next = nullptr;
+    if (__tail_ == nullptr)
+    {
+      __head_ = __item;
```

Comment:
Nitpick: checking for `__head_` feels appropriate here optionally we could just check for `empty()`